### PR TITLE
Write migration for changing sub categories

### PR DESCRIPTION
**Pull request by mohammadranjbarz**
related to https://github.com/Giveth/GIVeconomy/issues/723

### CODE REVIEW

**Comment by mohammadranjbarz** [comment]
The categories should change like 
https://docs.google.com/spreadsheets/d/1DAB5UHCYO4hkw_RLRtz8DLXpGduVDRMHfiW-71H0m8o/edit#gid=122470553

If you rould https://github.com/Giveth/GIVeconomy/issues/723 you would know

File: migration/1664367797442-changeMainCategoriesAndSubCategories.ts
Comment (CarlosQ96) [inline migration/1664367797442-changeMainCategoriesAndSubCategories.ts:28]: This query can be improved. There is not need to fetch all info or rows if you only need the id of the last record.
```
SELECT id
FROM main_category
ORDER BY id DESC
LIMIT 1
```

File: src/services/changeAPI/nonProfits.ts
Comment (CarlosQ96) [inline src/services/changeAPI/nonProfits.ts]: Move this error to the errorMessages.ts file so we keep them organized.

File: migration/1664367797442-changeMainCategoriesAndSubCategories.ts
Comment (CarlosQ96) [reply migration/1664367797442-changeMainCategoriesAndSubCategories.ts:28]: Is this query required?

File: migration/1664367797442-changeMainCategoriesAndSubCategories.ts
Comment (CarlosQ96) [inline migration/1664367797442-changeMainCategoriesAndSubCategories.ts:46]: Is it required to do ++lastId?
If you insert them in that order, each will have the id higher than the other. In fact I don't think you even need to know the last id, by default it will increase it with each subsequent insert in order.

That would remove the need to query the last id.

File: src/services/the-giving-blocks/syncProjectsCronJob.ts
Comment (CarlosQ96) [inline src/services/the-giving-blocks/syncProjectsCronJob.ts:171]: Validate with frontend if they are checking if the project has givingBlocksId or if they are checking category, this might break the logo they put on the project list if they are checking the category.

File: src/services/the-giving-blocks/syncProjectsCronJob.ts
Comment (CarlosQ96) [inline src/services/the-giving-blocks/syncProjectsCronJob.ts]: Move this to the errorMessages.ts so we keep them organized!

**Review by CarlosQ96 — CHANGES_REQUESTED** [review]
Hey Mohammad Thanks! I have some questions and improvements for the migration.
And some minor Stuff, but in general I think it looks good.

We can merge after resolving.

File: migration/1664367797442-changeMainCategoriesAndSubCategories.ts
Comment (mohammadranjbarz) [reply migration/1664367797442-changeMainCategoriesAndSubCategories.ts:28]: Thanks carlos , I changed it

File: migration/1664367797442-changeMainCategoriesAndSubCategories.ts
Comment (mohammadranjbarz) [reply migration/1664367797442-changeMainCategoriesAndSubCategories.ts:46]: Yeah at the beginning I didnt' pass the id, but I was getting this error, you can test it yourself in your system, I didn't know what was the reason but after passing the id it worked

Screen Shot 1401-07-10 at 12 10 58

https://github.com/Giveth/impact-graph/actions/runs/3151490462/jobs/5125501150

File: src/services/the-giving-blocks/syncProjectsCronJob.ts
Comment (mohammadranjbarz) [reply src/services/the-giving-blocks/syncProjectsCronJob.ts:171]: I think they check `givingBlocksId` but let ask them

@MohammadPCh 
How do you know a project is for giving blocks, from their category or `givingBlocksId` field in project entity?

File: src/services/the-giving-blocks/syncProjectsCronJob.ts
Comment (mohammadranjbarz) [reply src/services/the-giving-blocks/syncProjectsCronJob.ts]: Oh you are right, I moved that to errorMessages

File: src/services/changeAPI/nonProfits.ts
Comment (mohammadranjbarz) [reply src/services/changeAPI/nonProfits.ts]: Done

**Comment by mohammadranjbarz** [comment]
@CarlosQ96  Thank as lot for you review, I did the change requests

File: migration/1664367797442-changeMainCategoriesAndSubCategories.ts
Comment (CarlosQ96) [reply migration/1664367797442-changeMainCategoriesAndSubCategories.ts:46]: Let me do a quick test and get back to you. Thanks for mentioning this.

Comment (CarlosQ96) [reply migration/1664367797442-changeMainCategoriesAndSubCategories.ts:46]: This should actually work when creating the table, but it doesn't auto increment on insert
```SQL
id SERIAL NOT NULL
```
It could be a typeorm bug.
This is fixed later on when the server runs because we have typeorm synchronization.

Comment (CarlosQ96) [reply migration/1664367797442-changeMainCategoriesAndSubCategories.ts:46]: I tried to find a way to update the metadata of the column in the migration. But there is no easy fix. Let's leave it like this for now with the lastId.

For future migrations when creating tables we should add
```
isGenerated: true,
generationStrategy: 'increment',
```
with typeorm helpers.

**Review by CarlosQ96 — APPROVED** [review]
Thanks Mohammad!

**Comment by CarlosQ96** [comment]
Actually, upon further inspection since the column is type Serial it should auto increment. But it's not. It might be a typeorm bug 😢

**Comment by CarlosQ96** [comment]
Right now is fine as Synchronization fixes it when server run. I couldn't really find a fix at the moment. But let's not get stuck on this. I'll keep investigating later on.

**Comment by mohammadranjbarz** [comment]
@MoeNick  @maryjaf  Now it's ready to emerge , but I don't merge it now because the changes are huge and testing this issue might be a little complex, so I suggest first you pick some project in staging and note their categories, after merge see if the category of those projects has changed properly or not, so when you tell me, I would merge this

**Comment by maryjaf** [comment]
> Now it's ready to emerge , but I don't merge it now because the changes are huge and testing this issue might be a little complex, so I suggest first you pick some project in staging and note their categories, after merge see if the category of those projects has changed properly or not, so when you tell me, I would merge this

Thanks @mohammadranjbarz 
DONE

**Pull request merged**
